string.c

```diff
@@ -3109,14 +3109,15 @@ rb_str_resize(VALUE str, long len)
 
     int independent = str_independent(str);
     long slen = RSTRING_LEN(str);
+    const int termlen = TERM_LEN(str);
 
-    if (slen > len && ENC_CODERANGE(str) != ENC_CODERANGE_7BIT) {
+    if ((slen > len && ENC_CODERANGE(str) != ENC_CODERANGE_7BIT) ||
+        (termlen > 1 && (slen % termlen == 0) != (len % termlen == 0))) {
```

Comment:
I think you need to clear if `slen < len` (truncation) regardless.

I'd advocate for the simplest condition that catch the most common cases. I don't think the goal should be to never clear the coderange when it's theoretically avoidable, just to use fast and robust heuristics to avoid clearing it in the overwhelming majority of cases.

Reply:
> I'd advocate for the simplest condition

I'd like to use `slen != len && ENC_CODERANGE(str) != ENC_CODERANGE_7BIT` because it's simple and respects previous optimization for 7bit coderange. (not updated the pull request to this condition yet)

I think there are about four options.

```c
// 1
slen != len
// Most robust

// 2
slen != len && ENC_CODERANGE(str) != ENC_CODERANGE_7BIT
// ENC_CODERANGE_7BIT means it's ascii compatible encoding and all bytes are <=127,
// so truncate and extend will never change coderange.

// 3
(slen > len && ENC_CODERANGE(str) != ENC_CODERANGE_7BIT) || (slen < len && termlen > 1)
// Need non obvious knowledge that every character(codepoint = 1..0x10ffff) in
every termlen==1 encodings currently defined in `Encoding.list` does not end with "\0".

// 4
(slen > len && ENC_CODERANGE(str) != ENC_CODERANGE_7BIT) || (termlen > 1 && (slen % termlen == 0) != (len % termlen == 0))
// Most efficient
// Need additional knowledge that every termlen > 1 encodings (utf16 and utf32) are fixed-width,
// and the latter part of surrogate pair(utf16) is not "\0\0" so extending by 2bytes does not change broken to valid.
```

What do you think?

Reply:
I think the cases that are worth optimizing for are the very common encodings, namely: `ASCII`, `ASCII-BIT` and `UTF-8`, and in the increase case, which is common when appending strings.

Truncation is much rarer in my opinion, so not worth the complexity.

So I would advocate for:

```c
bool truncating = slen > len;
if (truncating || termlen != 1) {
  CLEAR_CODERANGE(...)
}
```

We lose the optimization for truncating ASCII-only strings, but I really doubt it's common enough to justify the complexity.

Of course I don't pretend to be an expert in this area, so I'm happy to defer to @nobu's opinion.

Reply:
👍